concrete: Implement Zeilberger's algorithm for finding recurrences

Kircheisser opened the pull request:
I have written an implementation of Zeilberger's algorithm, if this is of interest. It allows one to compute a great deal of definite hypergeometric sums that can't be done by for example gosper. 

For example, if we want to calculate sum of F(n, k) = (-1)^k * binomial(2n, k)^3 from k = 0 to 2n, this algorithm provides us a first order recurrence in n that F satisfies, this leads to a first order recurrence our sum satisfies. This recurrence is then quickly solved and sum found to be (-1)^n (3n)! / (n!)^3. 

<!-- BEGIN RELEASE NOTES -->
* concrete
    * Added an implementation of Zeilbergers algorithm for computing definite hypergeometric sum and         integrated this implementation into the methods that compute definite sums, this allows some new sums to be computed.
<!-- END RELEASE NOTES -->

#### Other comments
If this is of interest there are other more general things that can be done in the same vein.

Kircheisser committed:
concrete: Implement Zeilberger's algorithm for finding recurrences

This implements Zeilberger's algorithm for finding recurrences that
hypergeometric functions satisfy. Given a hypergeometric function
F(n, k), we find a function G and polynomials a_0, ... , a_J in n
such that

a_0F(n, k) + a_1F(n+1, k) + ... + a_JF(n+J, k) = G(n, k+1) - G(n, k)

This is of great benefit when finding definite sums that do not have
closed form expressions for their indefinite sums.

A simple example,

>>> F = binomial(n, k)
>>> zb_recur(F, 1, n, k)
([-2, 1], k/(k - n -1))

This tells us that for F(n, k) = binomial(n, k) we have
-2F(n, k) + F(n+1, k) = G(n, k+1) - G(n, k)
where G(n, k) = F * k/(k - n -1) (and 0 for k>n)

Now denote the sum of F(n, k) from k = 0 to n by f(n), summing
both sides of the recurrences from k = 0 to n+1, the right hand side
telescopes and we arrive at -2f(n) + f(n + 1) = 0. After checking an
initial condition we find f(n) = 2^n.

For another example taking
F(n, k) = (-1)^k * binomial(n, k) / binomial(x+k, k)
we quickly find the sum from k = 0 to n to be x / (x + n)

Kircheisser committed:
Add missing break in for loop

At some point in the algorithm we (clumsily) find the first symbol
in an expression, we intended to break out after finding it and failed
to do so.

Abdullahjavednesar commented on sympy/concrete/zeilberger.py:
```diff
@@ -0,0 +1,174 @@
+    return [c[1] for c in coeffs[0:J + 1]], combsimp(G / F)
+
+"""
+Tests
```
Tests should be kept separate.

Kircheisser committed:
Remove tests, to be later placed in test folder

Also changed formatting of zb_recur docstring as did not intend for
it to provide tests via ./bin/doctest.

Kircheisser committed:
Fix whitespace, end file with newline

asmeurer commented:
This needs tests.

asmeurer commented:
This is promising. The example you gave gives:

```
>>> pprint(summation((-1)**k * binomial(2*n, k)**3, (k, 0, 2*n)))
⎧         3
⎪⎛  2⋅n  ⎞   ┌─  ⎛        1, 1, 1, 1        │  ⎞    ┌─  ⎛-2⋅n, -2⋅n, -2⋅n │  ⎞
⎪⎜       ⎟ ⋅ ├─  ⎜                          │ 1⎟ +  ├─  ⎜                 │ 1⎟  for -6⋅n - 2 < 0
⎪⎝2⋅n + 1⎠  4╵ 3 ⎝2⋅n + 2, 2⋅n + 2, 2⋅n + 2 │  ⎠   3╵ 2 ⎝      1, 1       │  ⎠
⎪
⎪                              2⋅n
⎪                              ____
⎨                              ╲
⎪                               ╲              3
⎪                                ╲      k ⎛2⋅n⎞
⎪                                ╱  (-1) ⋅⎜   ⎟                                    otherwise
⎪                               ╱         ⎝ k ⎠
⎪                              ╱
⎪                              ‾‾‾‾
⎩                             k = 0
```

which looks like the right answer, except SymPy can't simplify it:

```py
>>> a = summation((-1)**k * binomial(2*n, k)**3, (k, 0, 2*n))
>>> b = (-1)**n*factorial(3*n)/factorial(n)**3
>>> [hyperexpand(a.subs(n, i)) == b.subs(n, i) for i in range(10)]
[True, True, True, True, True, True, True, True, True, True]
```

Have you looked into what it will take to include this in `Sum.doit`?

Kircheisser committed:
Add tests, both in docstring and separate file

There are 3 examples in the docstring of zb_recur. We check the
entire result of 2 and just the recurrence of 1. In a different file
we have a number of tests, for convenience we just check that
the function G/F returned by the function is correct. It is highly
unlikely this function be correct and the recurrence wrong.

Kircheisser commented:
I have looked into it briefly. It shouldn't be so hard. Certainly given the recurrence for F(n, k) it is simple to get to the sum itself. However it involves solving a recurrence relation with polynomial coefficients, e.g. via rsolve. 

This is fine for order one, but rsolve does not seem to produce the most general answers for higher order recurrences, which may lead to it not always finding a solution when it should.

asmeurer commented on sympy/concrete/zeilberger.py:
```diff
@@ -0,0 +1,144 @@
+    b_deg = _find_b_deg(p_2, p_3, p)
+
+    if b_deg < 0:
+        return "try higher order"
```
Functions should not return strings. If there is an error, it should raise an exception (like `raise ValueError("try higher order")`).

asmeurer commented on sympy/concrete/zeilberger.py:
```diff
@@ -0,0 +1,144 @@
+    if b_deg < 0:
+        return "try higher order"
+
+    b = Poly(sum(w(i) * k**i for i in range(b_deg + 1)), k)
```
You can just pass Poly a list of coefficients. It will be more efficient that way too.

Kircheisser committed:
Add function to compute sums via zb_recur

This function attempts to compute tricky definite sums by find a
recursion with zb_recur, this function presently has some issues
and is not doing as much as it could be. Nonetheless it allows one
to compute sums which sympy is seemingly unable to do (e.g. both
examples in the docstring).

An example,

F = k * binomial(2 * n + 1, 2 * k + 1), calling
zb_sum(F, n, 1, n, k) gives

4*2**(2*n - 3)*factorial(n - 1/2)/factorial(n - 3/2)

which is correct, and simplifies to 4^(n-1) * (2n - 1).

Presently zb_sum has some problems when it comes to actually making
use of the recurrence (which is working fine), for example rsolve
does not produce the correct result for the recurrence that
(-1)^k * binomial(2n, k)^3 satisfies and as a result can not arrive at
the sum.

Other changes -

Now zb_recur gives a value error rather than returning a string.

In zb_recur we at some point create a polynomial in k and we are
able to just provide the coefficients in a list, so we do that.

The docstring of zb_recur was slightly changed. Previously it
asserted that something it returned was a polynomial, while this
can be made to be true. There is no make it be so, the docstring
now better reflects reality.

Kircheisser commented:
I have added a function which attempts to do sums with zb_recur. It does some things well and others not so much, for sure it needs some thought.

An example lifted from the docstring.

>>> F = (-1)**k * binomial(x - k + 1, k) * binomial(x - 2 * k, n - k)
>>> zb_sum(F, n, 2, n, k)
(-1)**n/2 + 1/2

A big issue however is rsolve. It for example is not able to solve the recurrence for (-1)^k * binomial(2n,k)^3 (zb_recur for sure provides the correct recurrence). Many other examples exist (perhaps most that zb_recur will provide).

A particularly simple example of where this is an issue, we attempt to compute the sum of 
F(n, k) = binomial(n, k) * binomial(2k, k) / (-2)^k.

zb_recur informs us that the sum satisfies the recurrence f(n) - (n + 2)*f(n + 2)/(n + 1) = 0, which is correct. But rsolve is unable to solve it.

In fact, there is a first order recursion rsolve fails at

>>> r = n*f(n) + f(n+1)
>>> rsolve(r, f(n))
None

Such trivial recursions are likely quite common in these sorts of hypergeometric sums - there are a number of identities where things are say 1 for n = 0 and otherwise 0.

asmeurer commented:
So it sounds like we may need to improve rsolve before this can be really be useful. Even so, it may be useful to integrate this into Sum, even if it can only compute a subset of what it could because of rsolve.

Kircheisser committed:
Prepare zb_sum to be incorporated into Sum

Changed the arguments so that the function can be called just with
what is provided to sum. For now we will check conditions so that
vaniish is true and won't attempt to use zb_sum if it is not.

Also corrected some limits that would be wrong for more exotic
upper limits than n + c.

Kircheisser committed:
Add function to detect if a function vanishes

Now zb_sum can be called with just the information provided to it
by summation. So it can now conceptually be included. There is now
a function which tries to work out if a hypergeometric term vanishes
past some window by looking at the roots of F(n, k + 1) / F(n, k).

Noticed that thing were being called hypergeometric functions when
what was meant was hypergeometric term, so that's changed.

Kircheisser committed:
Assert that n is positive integer when doing substitution.

Need this so that it simplifies correctly.

Kircheisser committed:
Integrate zb_sum into sum

Now summation calls zb_sum if it has decided it has something
hypergeometric and is just about to return the piecewise representation.

We call zb_sum with J = 1 and J = 2 only to make it less likely the sum
gets caught in zb_sum for a long time. This still allows for a great
deal of sums, 'most' of the ones of interest that can be done are
liable to be caught just with J = 1, maybe.

Now for example summation(binomial(n, k) + 1, (k, 0, n)) will return
2**n + n + 1.

Will see if this breaks anything and if not have a clean up and check
everything is fine.

Kircheisser commented:
I have implemented zb_sum now it will definitely need more checking by me, anywho, I have broken some tests, I will need to see what's happened.

The first one I see 

def test_issue_2787():
    n, k = symbols('n k', positive=True, integer=True)
    p = symbols('p', positive=True)
    binomial_dist = binomial(n, k)*p**k*(1 - p)**(n - k)
    s = Sum(binomial_dist*k, (k, 0, n))
    res = s.doit().simplify()
    assert res == Piecewise(
        (n*p, p/Abs(p - 1) <= 1),
        ((-p + 1)**n*Sum(k*p**k*(-p + 1)**(-k)*binomial(n, k), (k, 0, n)),
        True))

It looks like zb_sum is correctly returning np, so this test is no longer a good test for whatever this issue is. I'll try and have a look later and replace it with a test that zb_sum doesn't 'fix'.

asmeurer commented:
If Zeilberger isn't able to give convergence conditions you might want to move it below the meijerg algorithms that do give them so that they are tried first.

Kircheisser commented:
Sorry which convergence do you mean? Zeilberger (I believe, I am somewhat learning as I go along) should be such that when it tries to sum say from 0 to n (I only say things about finite sums), the resulting expression will be the correct one for positive integer n where it's defined provided the summand is defined everywhere in the sum.

For the above binomial distribution example, do we not want it to return np, given that n is positive integer?

asmeurer commented:
Ah you're right. I don't know if the condition given for `p` is useful. The bounds given by Sum or Integral aren't guaranteed to be tight. 

If it only deals with finite sums I think having Zeilberger first should be fine.

Kircheisser committed:
Change zb_sum to only accept positive integer n

zb_sum is only valid for such n and summation allows for more than
this, so we need to check.

Changed a test in sums_products, previously a piecewise result
was being returned, while this was correct, zb_sum produces the
exact result.

Kircheisser committed:
Change summation to sum in zb_sum

Before zb_sum could potentially recursively call itself if there
were strange upper limits on the sum like n^2. If it couldn't solve
the resulting thing this could cause an infinite loop.

Kircheisser committed:
Tidy up zeilberger.py and change summation

Added try-catch to zb_sum as it seemed the neatest way to deal with
G_a or G_b not being defined, or rsolve recieving too complicated
an expression.

Summation used to call zb_sum with J=1 and J=2, but this can lead to
some sums taking quite a lot of time while zb_recur tries to solve the
J=2 case, and while often a recurrence is produced, rsolve is unable
to solve it so ultimately None is returned anyway.

Kircheisser committed:
Resubmitting to be tested

Last build errored but with no discernable error message and tests
all pass my side, checking to see if there really was just a random
error.

Kircheisser commented:
I think it should all be fine now, not sure why the commit before errored. Resubmitted the same thing and it worked this time.

Abdullahjavednesar commented:
Ping @asmeurer

asmeurer commented on sympy/concrete/summations.py:
```diff
@@ -1136,6 +1136,9 @@ def eval_sum_hyper(f, i_a_b):
             if res is not None:
                 return Piecewise(res, (old_sum, True))
         else:
+            zb_try = zb_sum(f, i_a_b)
```
The eval_sum_hyper function is specifically for the hypergeometric algorithm. This function should be put in in the eval_sum_symbolic function.

asmeurer commented on sympy/concrete/zeilberger.py:
```diff
@@ -0,0 +1,237 @@
+    """
+    Attempts to compute the sum of a hypergometric term F(n, k) from
+    k = a (finite) to k = b (finite) (which is assumed to be a function of n, a positive
+    integer, and nothing else) by first finding a recurrence of order J in n that
```
What are the complications from allowing symbolic constants?

asmeurer commented:
I have just the one comment and question.

Kircheisser commented:
Right I will change that. 

The complication is as so, we need to somehow infer n (the algorithm works for finding sums of a function f(n,k) over k by finding a recurrence of the sum in n) as it is not provided by the user (they just provide a function, the index to sum over and some limits, ideally I wanted to just look at the upper limit, see its a function of n and be done.

There is a way to do this without worrying about the limits, namely that you could look at the free symbols of the function, and check for each symbol whether it satisfied some suitable properties. I was worried that potentially this could make things slow. But now I think about it maybe it is not so. I will think about this.

Kircheisser committed:
Move zb_sum in summation, allow symbolic constants in zb_sum

Now zb_sum is in eval_sum_symbolic. By determining n in a different
way we now allow for symbolic constants in the limits of a sum.

jksuom commented:
It looks like Travis was heavily loaded and the execution time grew too long. I have restarted the job. It should pass with any luck.

Abdullahjavednesar commented:
Tests seem to pass. Ping @asmeurer @jksuom.

asmeurer commented on sympy/concrete/zeilberger.py:
```diff
@@ -192,13 +191,12 @@ def zb_sum(F, k_a_b, J = 1):
-        hypersimp(F, k) is None or hypersimp(F, n) is None):
+    for w in F.free_symbols.intersection(b.free_symbols.union(a.free_symbols)):
+        n = w
+        if  (n.is_positive and n.is_integer) and not (
```
It's generally an antipattern to require assumptions to be set. Instead just check that they aren't False and allow None.

asmeurer commented:
Sorry for not getting to this. I just have the one comment.

asmeurer commented:
Yes I mean it should work for a Symbol that has no assumptions set, like just `Symbol('n')`. You can return a Piecewise if you want, to keep the result mathematically valid.

Kircheisser committed:
Make zb_sum not require assumptions

Now zb_sum does not require assumptions to compute a sum, instead
it will pretend it has those assumptions, then in summation we will
return a piecewise function. This makes things slightly confusing to
read, we create a new symbol which plays the role of an old symbol
with additional assumptions, and have to substitute this in everywhere.

Kircheisser commented:
Made the changes now, not sure how happy I am with how it's done but it seems to work. For the definition of piecewise functions I couldn't find a way to say something is true for x in some generic set, without having say an indicator on this set and then declaring some equality? Maybe I just missed something.

Kircheisser committed:
Change zb_sum to deal with symbolic constant in limit better

Before there was a chance of zb_sum errantly picking up the wrong
symbol in the limit to use for a calculation. By only considering
symbols of b-a we remove some chance of this. Also _n has been
changed to n to read easier.

Abdullahjavednesar commented:
Ping @asmeurer

Kircheisser commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -886,11 +886,7 @@ def test_issue_2787():
-        ((-p + 1)**n*Sum(k*p**k*(-p + 1)**(-k)*binomial(n, k), (k, 0, n)),
-        True))
-
+    assert res == n*p
```
Hey,

Sorry I haven't looked at this in quite some time now. Nonetheless I can interpret this bit, 
It deals with the sum of binomial(n, k)*p**k*(1 - p)**(n - k) * k from k = 0 to n

The previous test asserts it to be np when p/|p-1| <= 1 and otherwise it was just returning the formal sum, this result is certainly correct but in fact this business of p/|p-1| <= 1 is needless, the sum is true for all real p. Presumably however the sum was being done prior it was doing something that just proves things on some restricted domain.

I think this is probably what a discussion between me and asmeurer a number of posts up concerned but I don't really remember.

Kircheisser commented:
Hey, just seen this, I'll add these release notes.

Commit:
Merge branch 'master' into zeilberger_algorithm

czgdp1807 commented:
@Kircheisser Are you still working on it? Please resolve the conflicts. Thanks.

czgdp1807 commented:
ping @smichr @asmeurer Can we continue this?

Chinrank commented:
Hello, I am Kircheisser (though I have since forgotten my credentials for that account). I completely forgot about this after I got a job as a software developer. I still understand what is going on and am happy to continue; and I would hope I can do a better job programming wise now, if there is still an interest?

asmeurer commented:
I think so, unless the algorithms have already been implemented. You can check some of the test cases in SymPy master. You'll need to start a new PR if you can't push to your old account.

sachin-4099 commented:
I will try to work on this along with @Chinrank to merge this PR.

Chinrank commented:
That's great @sachin-4099, I intend to look at it this weekend.

My belief is that we should go through what is already there, document it, confirm its correctness, and write it a bit more nicely. In the code there is a comment referencing where the algorithm can be found, the algorithm itself is reasonably complex but what is written already should be a straightforward implementation of it. Provided you are not already familar with this algorithm I think a good litmus test for this being a reasonable peice of work is that you are satisfied it is correct and understand it.

sachin-4099 commented:
@Chinrank, I am planning to do all the piled up or stalled PRs work this summer as my GSoC proposal.
This PR will come under that. No issue we will start working on this PR from this weekend onwards. It might get merged a bit late.

Commit:
Merge in sympy master

Commit:
Update failing test

Commit:
Resolve whitespace issue and unused import

Kircheisser commented:
I managed to login to my account. Looking through the code and rereading the algorithm, the implementation looks to me to be correct. @sachin-4099 let me know what you think and I can make further changes.

czgdp1807 commented:
Thanks @Kircheisser for the update. I have removed the `Please take over` label. Feel free to push more commits to this PR. Thanks for the contributions.

abhinav28071999 commented:
@Kircheisser This seems good addition to sympy. Are you still working on this? If not, then i would love to take over this pr.

czgdp1807 commented:
@abhinav28071999 Would you still like to take this over?
